File: Groove.SP/Groove.SP.Core/Entities/ShipmentModel.cs
using Groove.SP.Core.Events;
using Groove.SP.Core.Models;
using System;
using System.Collections.Generic;

namespace Groove.SP.Core.Entities
{
    public class ShipmentModel : Entity
    {
        public long Id { get; set; }

        public string ShipmentNo { get; set; }

        public string BookingNo { get; set; }

        public string BuyerCode { get; set; }

        public string CustomerReferenceNo { get; set; }

        public string AgentReferenceNo { get; set; }

        public string ShipperReferenceNo { get; set; }

        public string ModeOfTransport { get; set; }

        public DateTime CargoReadyDate { get; set; }

        public DateTime BookingDate { get; set; }

        public string ShipmentType { get; set; }

        public string ShipFrom { get; set; }

        public DateTime ShipFromETDDate { get; set; }

        public string ShipTo { get; set; }

        public DateTime? ShipToETADate { get; set; }

        public string Movement { get; set; }

        public int? Factor { get; set; }

        public decimal TotalPackage { get; set; }

        public string TotalPackageUOM { get; set; }

        public decimal TotalUnit { get; set; }

        public string TotalUnitUOM { get; set; }

        public decimal TotalGrossWeight { get; set; }

        public string TotalGrossWeightUOM { get; set; }

        public decimal TotalNetWeight { get; set; }

        public string TotalNetWeightUOM { get; set; }

        public decimal TotalVolume { get; set; }

        public string TotalVolumeUOM { get; set; }

        public string ServiceType { get; set; }

        public string Incoterm { get; set; }

        public string Status { get; set; }
        
        public bool IsFCL { get; set; }

        public bool IsItineraryConfirmed { get; set; }

        public long? POFulfillmentId { get; set; }

        public OrderType OrderType { get; set; }

        public string CommercialInvoiceNo { get; set; }

        public DateTime? InvoiceDate { get; set; }

        /// <summary>
        /// To map with <see cref="ContractMasterModel.CarrierContractNo"/> then show <see cref="ContractMasterModel.RealContractNo"/>
        /// </summary>
        public string CarrierContractNo { get; set; }

        public ContractMasterModel ContractMaster { get; set; }

        public POFulfillmentModel POFulfillment { get; set; }

        public virtual ICollection<ShipmentBillOfLadingModel> ShipmentBillOfLadings { get; set; }

        public virtual ICollection<ShipmentContactModel> Contacts { get; set; }

        public virtual ICollection<CargoDetailModel> CargoDetails { get; set; }

        public virtual ICollection<ShipmentLoadModel> ShipmentLoads { get; set; }

        public virtual ICollection<ShipmentLoadDetailModel> ShipmentLoadDetails { get; set; }

        /// <summary>
        /// To define linkage between shipment and master bill of lading
        /// </summary>
        public virtual ICollection<ConsignmentModel> Consignments { get; set; }

        /// <summary>
        /// To define linkage between shipment and master bill of lading
        /// </summary>
        public virtual ICollection<ConsignmentItineraryModel> ConsignmentItineraries { get; set; }

        public virtual ICollection<BillOfLadingConsignmentModel> BillOfLadingConsignments { get; set; }

        public virtual ICollection<ShipmentItemModel> ShipmentItems { get; set; }

        public virtual ICollection<POFulfillmentAllocatedOrderModel> POFulfillmentAllocatedOrders { get; set; }

        #region Methods
        public void SetCancelledStatus()
        {
            Status = StatusType.INACTIVE;
            AddDomainEvent(new ShipmentCancelledDomainEvent(this.Id));
        }
        #endregion
    }
}

File: Groove.SP/groove-sp-web/src/app/core/models/shipments/shipment.model.ts
import { ConsignmentModel } from 'src/app/features/consignment/models/consignment.model';
import { ConsolidationModel } from 'src/app/features/consolidation/models/consolidation.model';
import { ContractMasterModel } from 'src/app/features/shipment/master-bill-of-lading/models/contract-master-model';
import { ActivityModel } from '../activity/activity.model';
import { BaseModel } from '../base.model';
import { OrderType, POFulfillmentStageType } from '../enums/enums';
import { ShipmentBillOfLadingModel } from './shipment-bill-of-lading.model';
import { ShipmentContactModel } from './shipment-contact.model';
import { ShipmentItemModel } from './shipment-item.model';
import { ShipmentLoadDetailModel } from './shipment-load-detail.model';

export interface ShipmentModel extends BaseModel {
    id: number;
    shipmentNo: string;
    buyerCode: string;
    customerReferenceNo: string;
    modeOfTransport: string;
    cargoReadyDate: string;
    bookingDate: string;
    shipmentType: string;
    shipFrom: string;
    shipFromETDDate: string;
    shipTo: string;
    shipToETADate: string | null;
    movement: string;
    totalPackage: number | null;
    totalPackageUOM: string;
    totalUnit: number | null;
    totalUnitUOM: string;
    totalGrossWeight: number | null;
    totalGrossWeightUOM: string;
    totalNetWeight: number | null;
    totalNetWeightUOM: string;
    totalVolume: number | null;
    totalVolumeUOM: string;
    serviceType: string;
    incoterm: string;
    status: string;
    isFCL: boolean;
    fulfillmentId: number | null;
    fulfillmentNumber: string;
    bookingReferenceNo: string;
    fulfillmentStage: POFulfillmentStageType;
    isItineraryConfirmed: boolean;
    cyEmptyPickupTerminalCode: string | null;
    cyEmptyPickupTerminalDescription: string | null;
    cfsWarehouseCode: string | null;
    cfsWarehouseDescription: string | null;
    cyClosingDate: string | null;
    cfsClosingDate: string | null;
    billOfLadingNos: { item1: number; item2: string; }[];
    masterBillNos: { item1: number; item2: string; }[];
    bookingReferenceNumber: string;
    orderType: string;
    executionAgentName: string;
    shipperName: string;
    consigneeName: string;
    lastestActivity: string;
    isConfirmContainer: number;
    isConfirmConsolidation: number;
    latestMilestone: string;
    shipper: string;
    originAgent: string;
    destinationAgent: string;
    consignee: string;
    nominee: string;
    commercialInvoiceNo: string | null;
    invoiceDate: string | null;
    /**From buyer compliance. */
    enforceCommercialInvoiceFormat: boolean | null;
    contacts: ShipmentContactModel[];
    activities: ActivityModel[];
    shipmentBillOfLadings: ShipmentBillOfLadingModel[];
    shipmentItems: ShipmentItemModel[];
    consignments: ConsignmentModel[];
    shipmentLoadDetails: ShipmentLoadDetailModel[];
    consolidations: ConsolidationModel[];

    carrierContractNo?: string;
    /**
     * Value is from contractMaster.
     */
    carrierContractType?: string;

    /**
     * To contain information on contract.contractType
     */
    contractMaster?: ContractMasterModel;

    isException: boolean;

}

export interface ShipmentMilestoneModel {
    activityCode: string;
    title: string;
}